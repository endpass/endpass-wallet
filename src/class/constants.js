export const REQUEST_TIMEOUT_MSEC = 30000;

export default {
  REQUEST_TIMEOUT_MSEC,
};
