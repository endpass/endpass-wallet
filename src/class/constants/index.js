export { INPAGE_EVENT, INPAGE_ID_PREFIX } from './InpageProvider';

export const REQUEST_TIMEOUT_MSEC = 30000;

export const STORAGE_USER_META_KEY = 'endpass-user-meta';

export const AVAILABLE_USER_META_PROPS = ['activeAccount'];

export const PROXY_REQUEST_PREFIX = 'endpass-identity';

export const HARDWARE_DERIVIATION_PATH = `m/44'/60'/0'/0/`;

export default {
  REQUEST_TIMEOUT_MSEC,
  STORAGE_USER_META_KEY,
  AVAILABLE_USER_META_PROPS,
  PROXY_REQUEST_PREFIX,
  HARDWARE_DERIVIATION_PATH,
};
