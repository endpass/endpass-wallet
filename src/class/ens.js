import namehash from 'eth-ens-namehash';
import ABI from '@/abi/ens.json';
import { web3 } from '@/utils';

const registryAddresses = {
  // Mainnet
  '1': '0x314159265dd8dbb310642f98f50c066173c1259b',
  // Ropsten
  '3': '0x112234455c3a32fd11230c42e7bccd4a84e02010',
  // Rinkeby
  '4': '0xe7410170f87102DF0055eB195163A03B7F2Bff4A',
  // Ethereum Classic
  '61': '0xb96836a066ef81ea038280c733f833f69c23efde',
};

export default class ENSResolver {
  static async getAddress(name) {
    const { Contract, net } = web3.eth;
    const registryContract = new Contract(ABI);
    const netId = await net.getId();
    const node = namehash.hash(name);

    Object.assign(registryContract.options, {
      address: registryAddresses[netId],
    });

    const address = await registryContract.methods.resolver(node).call();

    if (address === '0x0000000000000000000000000000000000000000') {
      throw new Error("Name isn't resolvable");
    }

    return address;
  }
}
