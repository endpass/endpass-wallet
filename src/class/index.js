export { providerFactory } from './provider';
export { Transaction } from './Transaction';
export { default as EventEmitter } from './EventEmitter';
export { NotificationError } from './error';
export { Address } from './address';
export { Wallet } from './wallet';
export { ENSResolver } from './ens';
export { Token } from './Token';
export { default as ERC20Token } from './erc20';
