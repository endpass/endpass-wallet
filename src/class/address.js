import web3 from 'web3';
import { WALLET_TYPE, HARDWARE_WALLET_TYPE } from '@/constants';
import { TrezorWallet } from '@/class';

const { isAddress, hexToBytes, bytesToHex, toChecksumAddress } = web3.utils;

// A wallet based on an Ethereum address string, to be used in place of
// ethereumjs-wallet when only the address is available
export default class Address {
  constructor({ address, info }) {
    const addressString = `0x${address.replace(/^0x/, '')}`;

    if (!isAddress(addressString)) {
      throw new Error('Not a valid Ethereum address');
    }

    this.address = hexToBytes(addressString);
    this.isPublic = true;
    this._privKey = null;
    this._pubKey = null;
    this.privKey = null;
    this.pubKey = null;

    if (info) {
      this.info = info;

      const isHardware = Object.values(HARDWARE_WALLET_TYPE).includes(
        info.type,
      );

      if (isHardware) {
        this.isPublic = false;
      }

      switch (info.type) {
        case WALLET_TYPE.TREZOR:
          this.signStrategy = TrezorWallet;
          break;

        default:
          this.signStrategy = TrezorWallet;
          break;
      }
    }
  }

  getAddress() {
    return this.address;
  }

  getAddressString() {
    return bytesToHex(this.address);
  }

  getChecksumAddressString() {
    return toChecksumAddress(this.getAddressString());
  }

  // (Message) => Promise<SignedMessage>
  sign(message) {
    return this.signStrategy.sign(message, this.info.index);
  }

  // (Transaction) => Promise<SignedTrx>
  signTransaction(transaction) {
    return this.signStrategy.signTransaction(transaction, this.info.index);
  }
}
