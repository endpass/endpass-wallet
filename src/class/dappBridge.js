import { INPAGE_EVENT } from '@/constants';
import EventEmitter from './EventEmitter';

class DappBridge extends EventEmitter {
  constructor(...props) {
    super(props);

    this.requestHandler = null;
    this.setupEventsHandlers();
  }

  setupEventsHandlers() {
    this.on(INPAGE_EVENT.REQUEST, this.handleRequest);
    this.on(INPAGE_EVENT.SETTINGS, this.handleRequest);
  }

  setRequestHandler(handler) {
    this.requestHandler = handler;
  }

  handleRequest(payload) {
    if (this.requestHandler) {
      this.requestHandler(payload);
    }
  }

  emitResponse(payload) {
    this.emit(INPAGE_EVENT.RESPONSE, payload);
  }

  emitSettings(payload) {
    this.emit(INPAGE_EVENT.SETTINGS, payload);
  }
}

export default new DappBridge();
