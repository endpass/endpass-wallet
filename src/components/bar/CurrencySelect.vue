<template>
  <div class="provider-select">
    <div class="net-select">
      <vue-multiselect
        :options="currencies"
        :show-labels="false"
        :allow-empty="true"
        :value="activeCurrency"
        track-by="id"
        label="name"
        placeholder="Select currency"
        data-test="currency-select"
        @select="selectCurrency"
      />
    </div>
  </div>
</template>

<script>
import VueMultiselect from 'vue-multiselect';
import { mapActions, mapState } from 'vuex';

import { CURRENCIES } from '@/constants';

export default {
  name: 'CurrencySelect',

  data() {
    return {
      currencies: CURRENCIES,
    };
  },
  computed: {
    ...mapState({
      activeCurrency: state => state.web3.activeCurrency,
    }),
  },
  methods: {
    ...mapActions('web3', ['changeCurrency']),
    selectCurrency(currency) {
      this.changeCurrency({
        currencyId: currency.id,
      });
    },
  },
  components: {
    VueMultiselect,
  },
};
</script>


<style lang="scss">
.provider-select {
  .net-select {
    display: inline-block;
    max-width: 100%;
    position: relative;
    vertical-align: top;
  }
}
</style>

<style src="vue-multiselect/dist/vue-multiselect.min.css">
</style>
