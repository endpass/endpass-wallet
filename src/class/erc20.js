import Web3 from 'web3';
import erc20ABI from '@/abi/erc20.json';
import { Token } from '@/class';

// Service for functions related to ERC20 tokens
export class ERC20Token {
  // Accepts a web.js instance and the address of the token contract
  constructor(address) {
    this.address = address;

    this._contract = null;
    this._token = null;
  }

  // Returns the web3.Contract instance for this token
  getContract() {
    if (!this._contract) {
      this._contract = new Web3.eth.Contract(erc20ABI, this.address);
    }
    return this._contract;
  }

  // Returns the balance of this token at an address in wei
  async getBalance(address) {
    let contract = this.getContract();
    return contract.methods.balanceOf(address).call();
  }

  // Returns information about the token: name, symbol, decimals, totalSupply
  async getInfo() {
    const tokenInfo = {};
    let contract = this.getContract();
    const methods = ['name', 'decimals', 'symbol', 'totalSupply'].map(
      async item => {
        try {
          tokenInfo[item] = await contract.methods[item]().call();
        } catch (e) {
          tokenInfo[item] = undefined;
        }
      },
    );
    await Promise.all(methods);
    return tokenInfo;
  }

  // Returns a Token class object by fetching ERC20 token info from the blockchain
  async getToken() {
    // Return cached token info if it has been set
    if (this._token && this._token.symbol && this._token.decimals) {
      return this._token;
    }

    let tokenInfo = await this.getInfo();

    this._token = new Token({
      address: this.address,
      ...tokenInfo,
    });

    return this._token;
  }
}
