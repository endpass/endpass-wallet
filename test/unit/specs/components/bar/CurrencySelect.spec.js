import { shallowMount, createLocalVue } from '@vue/test-utils';
import { Network } from '@endpass/class';
import CurrencySelect from '@/components/bar/CurrencySelect';
import Vuex from 'vuex';
const localVue = createLocalVue();

localVue.use(Vuex);

describe('CurrencySelect', () => {
  let wrapper, actions;
  beforeEach(() => {
    actions = {
      changeCurrency: jest.fn(),
    };
    const store = new Vuex.Store({
      modules: {
        web3: {
          namespaced: true,
          state: {
            activeCurrency: Network.CURRENCIES[0],
          },
          actions,
        },
      },
    });

    wrapper = shallowMount(CurrencySelect, {
      localVue,
      store,
    });
  });

  describe('render', () => {
    it('should be a Vue component', () => {
      expect(wrapper.name()).toBe('CurrencySelect');
      expect(wrapper.isVueInstance()).toBeTruthy();
    });

    it('should render initial state of the component', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
