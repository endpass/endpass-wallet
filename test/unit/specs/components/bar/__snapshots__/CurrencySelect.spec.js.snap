// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencySelect render should render initial state of the component 1`] = `
<div
  class="provider-select"
>
  <div
    class="net-select"
  >
    <vue-multiselect
      allow-empty="true"
      label="name"
      options="[object Object],[object Object],[object Object]"
      placeholder="Select currency"
      track-by="id"
      value="[object Object]"
    />
  </div>
</div>
`;
